### PR TITLE
Streams

### PR DESCRIPTION
**Pull request by Vadifire**

### CODE REVIEW

File: src/main/java/tweeting/resources/FilterHomeTimelineResource.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    @GET
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    public Response getHomeTimeline(@QueryParam(FILTER_PARAM) Optional<String> keyword) {
+        try {
+            if (keyword.isPresent()) {
+                final List<Tweet> tweets = service.getHomeTimeline();
+                final List<Tweet> filteredTweets = tweets.stream()
+                        .filter(t -> t.getMessage().toLowerCase().contains(keyword.get().toLowerCase()))
+                        .collect(Collectors.toList());
```
Comment (andrewcchau) [inline src/main/java/tweeting/resources/FilterHomeTimelineResource.java]: Is this the cleanest way to do it?

File: src/main/java/tweeting/resources/FilterHomeTimelineResource.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+@Produces(MediaType.APPLICATION_JSON)
+
+public class FilterHomeTimelineResource {
+
+    private static final Logger logger = LoggerFactory.getLogger(FilterHomeTimelineResource.class);
+
+    public static final String FILTER_PARAM = "keyword";
+
+    public static final String MISSING_FILTER_MESSAGE = "\"" + FILTER_PARAM + "\" parameter is required to filter" +
+            "home timeline. ";
```
Comment (andrewcchau) [inline src/main/java/tweeting/resources/FilterHomeTimelineResource.java]: Check your resulting string

File: src/main/java/tweeting/resources/TweetResource.java [hunk 3]
```diff
@@ -34,25 +34,29 @@ public PostTweetResource(TwitterService service) {
      */
     @POST
     @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
-    public Response postTweet(@FormParam("message") String message) { // Receives message from JSON data
+    public Response postTweet(@FormParam("message") Optional<String> message) {
```
Comment (kev-chan) [inline src/main/java/tweeting/resources/TweetResource.java]: you don't need optionals for the param

**Review by kev-chan — CHANGES_REQUESTED** [review]
keep looking at what you were looking at earlier, you haven't found the right thing yet.

File: src/main/java/tweeting/models/Tweet.java [hunk 4]
```diff
@@ -1,31 +1,32 @@
...
 
 public class Tweet {
 
     private String message;
     private TwitterUser user;
     private Date createdAt;
 
-    public String getMessage() {
-        return message;
+    public Optional<String> getMessage() {
```
Comment (kev-chan) [inline src/main/java/tweeting/models/Tweet.java]: you don't need optionals in the java objects

File: src/main/java/tweeting/services/TwitterService.java [hunk 5]
```diff
@@ -61,25 +63,29 @@ public static TwitterService getInstance() {
         return instance;
     }
 
-    public List<Tweet> getHomeTimeline() throws TwitterServiceResponseException {
+    public Optional<List<Tweet>> getHomeTimeline() throws TwitterServiceResponseException {
         try {
+            logger.info("Successfully retrieved home timeline from Twitter.");
```
Comment (kev-chan) [inline src/main/java/tweeting/services/TwitterService.java]: stream syntax here

File: src/main/java/tweeting/services/TwitterService.java [hunk 6]
```diff
@@ -61,25 +63,29 @@ public static TwitterService getInstance() {
...
             return constructTweetList(api.getHomeTimeline());
         } catch (TwitterException te) {
             throw createServerException(te);
         }
     }
 
-    public Tweet postTweet(String message) throws TwitterServiceResponseException, TwitterServiceCallException {
-        // Prelim checks (avoid calling to Twitter if unnecessary)
-        if (message == null) {
+    public Optional<Tweet> postTweet(String message) throws TwitterServiceResponseException,
```
Comment (kev-chan) [inline src/main/java/tweeting/services/TwitterService.java]: same as above

File: src/main/java/tweeting/resources/FilterHomeTimelineResource.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    /*
+     * How to use:
+     * curl -i -X GET http://HOST:PORT/api/1.0/tweet/filter?keyword=KEYWORD
+     *
+     * Replace 'KEYWORD' with desired keyword to filter by. Replace HOST and PORT with configured values
+     */
+    @GET
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getHomeTimeline(@QueryParam(FILTER_PARAM) String keyword) {
```
Comment (kev-chan) [inline src/main/java/tweeting/resources/FilterHomeTimelineResource.java]: how re-usable is this?

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/tweeting/resources/FilterHomeTimelineResource.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+    /*
+     * How to use:
+     * curl -i -X GET http://HOST:PORT/api/1.0/tweet/filter?keyword=KEYWORD
+     *
+     * Replace 'KEYWORD' with desired keyword to filter by. Replace HOST and PORT with configured values
+     */
+    @GET
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    public Response getHomeTimeline(@QueryParam(FILTER_PARAM) Optional<String> keyword) {
```
Comment (kev-chan) [inline src/main/java/tweeting/resources/FilterHomeTimelineResource.java]: was this just copy and pasted?

File: src/main/java/tweeting/resources/FilterHomeTimelineResource.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    /*
+     * How to use:
+     * curl -i -X GET http://HOST:PORT/api/1.0/tweet/filter?keyword=KEYWORD
+     *
+     * Replace 'KEYWORD' with desired keyword to filter by. Replace HOST and PORT with configured values
+     */
+    @GET
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getHomeTimeline(@QueryParam("keyword") String keyword) {
```
Comment (kev-chan) [inline src/main/java/tweeting/resources/FilterHomeTimelineResource.java]: was this just copy and pasted?

File: src/main/java/tweeting/services/TwitterService.java [hunk 10]
```diff
@@ -61,28 +65,48 @@ public static TwitterService getInstance() {
...
+            tweets.ifPresent(t -> logger.info("Successfully retrieved home timeline."));
+            return tweets;
         } catch (TwitterException te) {
             throw createServerException(te);
         }
     }
 
+    public Optional<List<Tweet>> getFilteredTimeline(String keyword) throws TwitterServiceResponseException,
+            TwitterServiceCallException {
+        if (keyword == null) {
```
Comment (kev-chan) [inline src/main/java/tweeting/services/TwitterService.java]: does this cover all of the cases?

File: src/main/java/tweeting/services/TwitterService.java [hunk 11]
```diff
@@ -92,28 +116,28 @@ private TwitterServiceResponseException createServerException(TwitterException t
...
     }
 
-    private LinkedList<Tweet> constructTweetList(List<Status> statuses) {
-        LinkedList<Tweet> listOfTweets = new LinkedList<>();
-        for (Status status : statuses) {
-            listOfTweets.add(constructTweet(status));
-        }
-        return listOfTweets;
+    private Optional<List<Tweet>> constructTweetList(List<Status> statuses) {
+        return Optional.of(statuses.stream()
```
Comment (kev-chan) [inline src/main/java/tweeting/services/TwitterService.java]: are all of these lines necessary?

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/tweeting/resources/FilterHomeTimelineResource.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    @GET
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getHomeTimeline(@QueryParam(FILTER_PARAM) String keyword) {
+        try {
+            if (keyword != null) {
+                return Response.ok(service.getHomeTimeline()
+                        .orElseThrow(() -> new NullPointerException("Twitter failed to respond with home timeline."))
+                        .stream()
+                        .filter(t -> t.getMessage().isPresent())
```
Comment (kev-chan) [inline src/main/java/tweeting/resources/FilterHomeTimelineResource.java]: what is this filter for?

File: src/main/java/tweeting/resources/TwitterResource.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    @Path("tweet/filter")
+    @GET
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getFilteredHomeTimeline(@QueryParam("keyword") String keyword) {
+        try {
+            return service.getFilteredTimeline(keyword)
+                    .map(timeline -> Response.ok(timeline)
+                            .build())
+                    .orElseThrow(() -> new NullPointerException("Twitter failed to respond with home timeline."));
```
Comment (kev-chan) [inline src/main/java/tweeting/resources/TwitterResource.java]: this feels odd to just throw it below, if this is the case, why not just have the other class throw the exception?

**Review by kev-chan — CHANGES_REQUESTED** [review]

**Pull request merged**